### PR TITLE
Fix gpio_sam.c ACCESS_BY_PORT mode

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
1. drivers: gpio_sam: Add ACCESS_BY_PORT mode for gpio config
2. drivers: gpio_sam: fix gpio_sam_write ACCESS_BY_PORT mode

### CODE REVIEW

**Comment by qianfan-Zhao** [comment]
@rgundi @fallrisk Could you please review this patch?

File: drivers/gpio/gpio_sam.c [hunk 1]
```diff
@@ -116,6 +107,32 @@ static int gpio_sam_config(struct device *dev, int access_op, u32_t pin,
...
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Pio *const pio = cfg->regs;
+	int i, result;
+
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		gpio_sam_config_pin(pio, BIT(pin), flags);
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		for (i = 0; i < 32; i++) {
```
Comment (rgundi) [inline drivers/gpio/gpio_sam.c]: Any reason why all the pins need to be configured even if only one of them is being requested? A comment over here can provide clarity.

**Review by rgundi — COMMENTED** [review]
The commit message can probably read "Configure all the pins for ACCESS_BY_PORT mode the same way it is done for ACCESS_BY_PIN".
Also, i don't see any changes for gpio_sam_read. That is also needed, right?
Your patchset also has checkpatch errors. Please check the shippable and fix it.

**Comment by qianfan-Zhao** [comment]
@rgundi 
> Also, i don't see any changes for gpio_sam_read. That is also needed, right?

The previous code already do that.

```
static int gpio_sam_read(struct device *dev, int access_op, u32_t pin,
			 u32_t *value)
{
	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
	Pio *const pio = cfg->regs;

	*value = pio->PIO_PDSR;

	switch (access_op) {
	case GPIO_ACCESS_BY_PIN:
		*value = (*value >> pin) & 1;
		break;
	case GPIO_ACCESS_BY_PORT:
		break;
	default:
		return -ENOTSUP;
	}

	return 0;
}
```

GPIO_ACCESS_BY_PORT mode save port's value to param 'value' and return 0 means successful.

File: drivers/gpio/gpio_sam.c (hunk 1 above)
Comment (qianfan-Zhao) [reply drivers/gpio/gpio_sam.c]: gpio_port_configure set all the pins the same way in the port, and others soc gpio drivers such as gpio_mcux.c implement thats. 

But I think it's no reason set all gpios in the same port to a same way, I opened an issue (#10305) to discuss this. I want add another param 'mask' to gpio_port_configure, only configure which pin marked in param 'mask'.

Could we discuss more details on issue #10305 about gpio_port_configure?
```
/**
 * @brief Configure all the pins the same way in the port.
 *        List out all flags on the detailed description.
 *
 * @param port Pointer to the device structure for the driver instance.
 * @param flags Flags for the port configuration. IN/OUT, interrupt ...
 * @return 0 if successful, negative errno code on failure.
 */
static inline int gpio_port_configure(struct device *port, int flags)
```

**Review by nashif — APPROVED** [review]
the API proposed in #10305 is ok, we should move to that.

**Pull request merged**